### PR TITLE
Ring buffer

### PR DESCRIPTION
**Pull request by dustinhiatt-wf**
CODE REVIEW

We have buffers in our codebase that are protected with locks and waitgroups for blocking.  We need a faster buffer and this is that.  This is an MPMC threadsafe ring buffer where threadsafety is achieved using only CAS operations.  Makes this buffer pretty fast, and the fixed size ensures bounded sorts of behavior.  Calling dispose on this buffer will free any blocked threads with an error returned (unlike panicking with channels). 

I'm hoping to use this to speed up the PALM tree. 

BenchmarkRBLifeCycle-8	10000000	       177 ns/op // involves simultaneous put/get of b.N items
BenchmarkRBPut-8	30000000	        58.1 ns/op
BenchmarkRBGet-8	50000000	        26.8 ns/op

@alexandercampbell-wf @beaulyddon-wf @tannermiller-wf @rosshendrickson-wf @tylertreat-wf @ericolson-wf @stevenosborne-wf

### CODE REVIEW

**Comment by tannermiller-wf** [comment]
+1 lockless datastructures are very cool

**Comment by tylertreat-wf** [comment]
+1

**Comment by beaulyddon-wf** [comment]
Awesome. +1

File: queue/queue.go [hunk 1]
```diff
@@ -24,10 +24,31 @@ as opposed to panicking as with channels.  Queues will grow with unbounded
...
+Benchmarks:
+BenchmarkPriorityQueue-8	 		2000000	       782 ns/op
+BenchmarkQueue-8	 		 		2000000	       671 ns/op
+BenchmarkChannel-8	 		 		1000000	      2083 ns/op
+BenchmarkQueuePut-8	   		   		20000	     84299 ns/op
+BenchmarkQueueGet-8	   		   		20000	     80753 ns/op
+BenchmarkExecuteInParallel-8	    20000	     68891 ns/op
+BenchmarkRBLifeCycle-8				10000000	       177 ns/op
+BenchmarkRBPut-8					30000000	        58.1 ns/op
+BenchmarkRBGet-8					50000000	        26.8 ns/op
```
Comment (stevenosborne-wf) [inline queue/queue.go]: Alignment is a little off here.

**Comment by stevenosborne-wf** [comment]
Super minor. +1

File: queue/queue.go (hunk 1 above)
Comment (tylertreat-wf) [inline queue/queue.go]: The benchmarking does this for some reason. Pretty annoying.

**Comment by alexandercampbell-wf** [comment]
+1

**Pull request merged**